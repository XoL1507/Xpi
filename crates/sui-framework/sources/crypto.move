// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

/// Library for cryptography onchain.
module sui::crypto {
    friend sui::validator;

    /// @param signature: A 65-bytes signature in form (r, s, v) that is signed using 
    /// Secp256k1. Reference implementation on signature generation using RFC6979: 
    /// https://github.com/MystenLabs/narwhal/blob/5d6f6df8ccee94446ff88786c0dbbc98be7cfc09/crypto/src/secp256k1.rs
    /// The accepted v values are {0, 1, 2, 3}.
    /// 
    /// @param hashed_msg: the hashed 32-bytes message. The message must be hashed instead 
    /// of plain text to be secure.
    /// 
    /// If the signature is valid, return the corresponding recovered Secpk256k1 public 
    /// key, otherwise throw error. This is similar to ecrecover in Ethereum, can only be 
    /// applied to Secp256k1 signatures.
    public native fun ecrecover(signature: vector<u8>, hashed_msg: vector<u8>): vector<u8>;

    /// @param data: arbitrary bytes data to hash
    /// Hash the input bytes using keccak256 and returns 32 bytes.
    public native fun keccak256(data: vector<u8>): vector<u8>;

    /// @param signature: A 48-bytes signature that is a point on the G1 subgroup
    /// @param public_key: A 96-bytes public key that is a point on the G2 subgroup
    /// @param msg: The message that we test the signature against.
    ///
    /// If the signature is a valid BLS12381 signature of the message and public key, return true.
    /// Otherwise, return false.
    public native fun bls12381_verify_g1_sig(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>): bool; 

    /// @param signature: A 65-bytes signature in form (r, s, v) that is signed using 
    /// Secp256k1. Reference implementation on signature generation using RFC6979: 
    /// https://github.com/MystenLabs/narwhal/blob/5d6f6df8ccee94446ff88786c0dbbc98be7cfc09/crypto/src/secp256k1.rs
    /// 
    /// @param public_key: The public key to verify the signature against
    /// @param hashed_msg: The hashed 32-bytes message, same as what the signature is signed against.
    /// 
    /// If the signature is valid to the pubkey and hashed message, return true. Else false.
    public native fun secp256k1_verify(signature: vector<u8>, public_key: vector<u8>, hashed_msg: vector<u8>): bool;

    use sui::elliptic_curve::{Self as ec, RistrettoPoint};

    /// Only bit_length = 64, 32, 16, 8 will work.
    native fun native_verify_full_range_proof(proof: vector<u8>, commitment: vector<u8>, bit_length: u64);

    /// @param proof: The bulletproof
    /// @param commitment: The commitment which we are trying to verify the range proof for
    /// @param bit_length: The bit length that we prove the committed value is whithin. Note that bit_length must be either 64, 32, 16, or 8.
    /// 
    /// If the range proof is valid, execution succeeds, else panics.
    public fun verify_full_range_proof(proof: vector<u8>, commitment: RistrettoPoint, bit_length: u64) {
        native_verify_full_range_proof(proof, ec::bytes(&commitment), bit_length)
    }

    /// @param signature: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param public_key: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param msg: The message that we test the signature against.
    /// 
    /// If the signature is a valid Ed25519 signature of the message and public key, return true.
    /// Otherwise, return false.
    public(friend) native fun ed25519_verify(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>): bool;

    /// @param signature: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param public_key: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param msg: The message that we test the signature against.
    /// @param domain: The domain that the signature is tested again. We essentially prepend this to the message.
    /// 
    /// If the signature is a valid Ed25519 signature of the message and public key, return true.
    /// Otherwise, return false.
    public(friend) fun ed25519_verify_with_domain(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>, domain: vector<u8>): bool {
        std::vector::append(&mut domain, msg);
        ed25519_verify(signature, public_key, domain)
    }
}
