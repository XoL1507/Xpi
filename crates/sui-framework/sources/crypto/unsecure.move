// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

module sui::unsecure {
    friend sui::validator;

    public native fun unsecure_verify(signature: &vector<u8>, public_key: &vector<u8>, msg: &vector<u8>): bool;

    public(friend) fun unsecure_verify_with_domain(
        signature: &vector<u8>,
        public_key: &vector<u8>,
        msg: vector<u8>,
        domain: vector<u8>
    ): bool {
        std::vector::append(&mut domain, msg);
        unsecure_verify(signature, public_key, &domain)
    }

}
