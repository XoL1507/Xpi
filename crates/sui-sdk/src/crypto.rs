// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use rand::{rngs::StdRng, SeedableRng};
use serde::{Deserialize, Serialize};
use signature::Signer;
use std::collections::BTreeMap;
use std::fmt::Write;
use std::fmt::{Display, Formatter};
use std::fs;
use std::fs::File;
use std::io::BufReader;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use bip39::Mnemonic;
use rand::rngs::adapter::ReadRng;

use sui_types::base_types::SuiAddress;
use sui_types::crypto::{
    get_key_pair_from_rng, EncodeDecodeBase64, PublicKey, Signature, SuiKeyPair,
};

#[derive(Serialize, Deserialize)]
#[non_exhaustive]
// This will work on user signatures, but not suitable for authority signatures.
pub enum KeystoreType {
    File(PathBuf),
    InMem(usize),
}

pub trait AccountKeystore: Send + Sync {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error>;
    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error>;
    fn keys(&self) -> Vec<PublicKey>;
}

impl KeystoreType {
    pub fn init(&self) -> Result<SuiKeystore, anyhow::Error> {
        Ok(match self {
            KeystoreType::File(path) => SuiKeystore::from(FileBasedKeystore::load_or_create(path)?),
            KeystoreType::InMem(initial_key_number) => {
                SuiKeystore::from(InMemKeystore::new(*initial_key_number))
            }
        })
    }
}

impl Display for KeystoreType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut writer = String::new();
        match self {
            KeystoreType::File(path) => {
                writeln!(writer, "Keystore Type : File")?;
                write!(writer, "Keystore Path : {:?}", path)?;
                write!(f, "{}", writer)
            }
            KeystoreType::InMem(_) => {
                writeln!(writer, "Keystore Type : InMem")?;
                write!(f, "{}", writer)
            }
        }
    }
}

#[derive(Default)]
pub struct FileBasedKeystore {
    keys: BTreeMap<SuiAddress, SuiKeyPair>,
    path: Option<PathBuf>,
}

impl AccountKeystore for FileBasedKeystore {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.keys
            .get(address)
            .ok_or_else(|| {
                signature::Error::from_source(format!("Cannot find key for address: [{address}]"))
            })?
            .try_sign(msg)
    }

    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        let address: SuiAddress = (&keypair.public()).into();
        self.keys.insert(address, keypair);
        self.save()?;
        Ok(())
    }

    fn keys(&self) -> Vec<PublicKey> {
        self.keys.values().map(|key| key.public()).collect()
    }
}

impl FileBasedKeystore {
    pub fn load_or_create(path: &Path) -> Result<Self, anyhow::Error> {
        let keys = if path.exists() {
            let reader = BufReader::new(File::open(path)?);
            let kp_strings: Vec<String> = serde_json::from_reader(reader)?;
            kp_strings
                .iter()
                .map(|kpstr| {
                    let key = SuiKeyPair::decode_base64(kpstr);
                    key.map(|k| (Into::<SuiAddress>::into(&k.public()), k))
                })
                .collect::<Result<BTreeMap<_, _>, _>>()
                .map_err(|_| anyhow::anyhow!("Invalid Keypair file"))?
        } else {
            BTreeMap::new()
        };

        Ok(Self {
            keys,
            path: Some(path.to_path_buf()),
        })
    }

    pub fn set_path(&mut self, path: &Path) {
        self.path = Some(path.to_path_buf());
    }

    pub fn save(&self) -> Result<(), anyhow::Error> {
        if let Some(path) = &self.path {
            let store = serde_json::to_string_pretty(
                &self
                    .keys
                    .values()
                    .map(EncodeDecodeBase64::encode_base64)
                    .collect::<Vec<_>>(),
            )
            .unwrap();
            fs::write(path, store)?
        }
        Ok(())
    }

    pub fn key_pairs(&self) -> Vec<&SuiKeyPair> {
        self.keys.values().collect()
    }
}

pub struct SuiKeystore(Box<dyn AccountKeystore>);

impl SuiKeystore {
    fn from<S: AccountKeystore + 'static>(keystore: S) -> Self {
        Self(Box::new(keystore))
    }

    pub fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        self.0.add_key(keypair)
    }

    pub fn generate_new_key(&mut self) -> Result<(SuiAddress, String), anyhow::Error> {
        let mnemonic = Mnemonic::generate(12)?;
        let seed = mnemonic.to_seed("");
        let mut rng = RngWrapper(ReadRng::new(&seed));
        // TODO(joyqvq): add ability to generate as Secp256k1 keypair
        let (address, kp) = get_key_pair_from_rng(&mut rng);
        self.0.add_key(SuiKeyPair::Ed25519SuiKeyPair(kp))?;
        Ok((address, mnemonic.to_string()))
    }

    pub fn keys(&self) -> Vec<PublicKey> {
        self.0.keys()
    }

    pub fn addresses(&self) -> Vec<SuiAddress> {
        self.keys().iter().map(|k| k.into()).collect()
    }

    pub fn signer(&self, signer: SuiAddress) -> impl Signer<Signature> + '_ {
        KeystoreSigner::new(&*self.0, signer)
    }

    pub fn import_from_mnemonic(&mut self, phrase: &str) -> Result<SuiAddress, anyhow::Error> {
        let seed = &Mnemonic::from_str(phrase).unwrap().to_seed("");
        let mut rng = RngWrapper(ReadRng::new(seed));
        let (address, kp) = get_key_pair_from_rng(&mut rng);
        // TODO(joyqvq): add ability to import as Secp256k1 keypair
        self.0.add_key(SuiKeyPair::Ed25519SuiKeyPair(kp))?;
        Ok(address)
    }

    pub fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.0.sign(address, msg)
    }
}

/// wrapper for adding CryptoRng and RngCore impl to ReadRng.
struct RngWrapper<'a>(ReadRng<&'a [u8]>);

impl rand::CryptoRng for RngWrapper<'_> {}
impl rand::RngCore for RngWrapper<'_> {
    fn next_u32(&mut self) -> u32 {
        self.0.next_u32()
    }

    fn next_u64(&mut self) -> u64 {
        self.0.next_u64()
    }

    fn fill_bytes(&mut self, dest: &mut [u8]) {
        self.0.fill_bytes(dest)
    }

    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
        self.0.try_fill_bytes(dest)
    }
}

struct KeystoreSigner<'a> {
    keystore: &'a dyn AccountKeystore,
    address: SuiAddress,
}

impl<'a> KeystoreSigner<'a> {
    pub fn new(keystore: &'a dyn AccountKeystore, account: SuiAddress) -> Self {
        Self {
            keystore,
            address: account,
        }
    }
}

impl Signer<Signature> for KeystoreSigner<'_> {
    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.keystore.sign(&self.address, msg)
    }
}

#[derive(Default)]
struct InMemKeystore {
    keys: BTreeMap<SuiAddress, SuiKeyPair>,
}

impl AccountKeystore for InMemKeystore {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.keys
            .get(address)
            .ok_or_else(|| {
                signature::Error::from_source(format!("Cannot find key for address: [{address}]"))
            })?
            .try_sign(msg)
    }

    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        let address: SuiAddress = (&keypair.public()).into();
        self.keys.insert(address, keypair);
        Ok(())
    }

    fn keys(&self) -> Vec<PublicKey> {
        self.keys.values().map(|key| key.public()).collect()
    }
}

impl InMemKeystore {
    pub fn new(initial_key_number: usize) -> Self {
        let mut rng = StdRng::from_seed([0; 32]);
        let keys = (0..initial_key_number)
            .map(|_| get_key_pair_from_rng(&mut rng))
            .map(|(ad, k)| (ad, SuiKeyPair::Ed25519SuiKeyPair(k)))
            .collect::<BTreeMap<SuiAddress, SuiKeyPair>>();

        Self { keys }
    }
}

impl AccountKeystore for Box<dyn AccountKeystore> {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        (**self).sign(address, msg)
    }

    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        (**self).add_key(keypair)
    }

    fn keys(&self) -> Vec<PublicKey> {
        (**self).keys()
    }
}
