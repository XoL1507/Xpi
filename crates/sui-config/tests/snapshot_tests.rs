// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

// This file contains tests that detect changes in Sui configs.
// If a PR breaks one or more tests here, the PR probably has a real impact
// on a production configuration file. When test failure happens, the PR should
// be marked as a breaking change and reviewers should be aware of this.
//
// Owners and operators of production configuration files can add themselves to
// .github/CODEOWNERS for the corresponding snapshot tests, so they can get notified
// of changes. PRs that modifies snapshot files should wait for reviews from
// code owners (if any) before merging.
//
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta
// 1. Run `cargo insta test --review` under `./sui-config`.
// 2. Review, accept or reject changes.

use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use std::path::PathBuf;

use insta::assert_yaml_snapshot;
use multiaddr::Multiaddr;
use narwhal_crypto::traits::KeyPair;
use rand::rngs::StdRng;
use rand::SeedableRng;
use sui_config::{genesis::Builder, genesis_config::GenesisConfig};
use sui_config::{NetworkConfig, ValidatorInfo};
use sui_types::base_types::{ObjectID, SuiAddress};
use sui_types::crypto::{get_key_pair_from_rng, AuthorityKeyPair, AuthorityPublicKeyBytes};

#[test]
fn genesis_config_snapshot_matches() {
    // Test creating fake SuiAddress from PublicKeyBytes.
    let keypair: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let public_key = AuthorityPublicKeyBytes::from(keypair.public());
    let fake_addr = SuiAddress::from(&public_key);

    let fake_obj_id = ObjectID::from(fake_addr);
    let mut genesis_config = GenesisConfig::for_local_testing();
    for account in &mut genesis_config.accounts {
        account.address = Some(fake_addr);
        for gas_obj in &mut account.gas_objects {
            gas_obj.object_id = fake_obj_id;
        }
    }
    assert_yaml_snapshot!(genesis_config);
}

// If a Move change breaks the test below, it is because related Move functions are included
// in the genesis blob. The test failure can be fixed by updating the snapshot / blob.
// Also, folks in the file's CODEWOWNERS should be notified about the change. They may need to
// re-create the genesis blob.
#[test]
fn empty_genesis_snapshot_matches() {
    let genesis = Builder::new().build();
    assert_yaml_snapshot!(genesis);
}

#[test]
fn populated_genesis_snapshot_matches() {
    let genesis_config = GenesisConfig::for_local_testing();
    let (_account_keys, objects) = genesis_config
        .generate_accounts(&mut StdRng::from_seed([0; 32]))
        .unwrap();
    let key: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let validator = ValidatorInfo {
        name: "0".into(),
        public_key: key.public().into(),
        stake: 1,
        delegation: 0,
        gas_price: 1,
        network_address: Multiaddr::empty(),
        narwhal_primary_to_primary: Multiaddr::empty(),
        narwhal_worker_to_primary: Multiaddr::empty(),
        narwhal_primary_to_worker: Multiaddr::empty(),
        narwhal_worker_to_worker: Multiaddr::empty(),
        narwhal_consensus_address: Multiaddr::empty(),
    };

    let genesis = Builder::new()
        .add_objects(objects)
        .add_validator(validator)
        .build();
    assert_yaml_snapshot!(genesis.validator_set());
    assert_yaml_snapshot!(genesis.committee().unwrap());
    assert_yaml_snapshot!(genesis.narwhal_committee());
    assert_yaml_snapshot!(genesis.sui_system_object());
    // Serialized `genesis` is not static and cannot be snapshot tested.
}

#[test]
fn network_config_snapshot_matches() {
    let temp_dir = tempfile::tempdir().unwrap();
    let committee_size = 7;
    let rng = StdRng::from_seed([0; 32]);
    let mut network_config = NetworkConfig::generate_with_rng(temp_dir.path(), committee_size, rng);
    // TODO: Inject static temp path and port numbers, instead of clearing them.
    for mut validator_config in &mut network_config.validator_configs {
        validator_config.db_path = PathBuf::from("/tmp/foo/");
        validator_config.network_address = Multiaddr::empty();
        let fake_socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 1);
        validator_config.json_rpc_address = fake_socket;
        validator_config.metrics_address = fake_socket;
        validator_config.admin_interface_port = 8888;
        let metrics_addr: Multiaddr = "/ip4/127.0.0.1/tcp/1234".parse().unwrap();
        if let Some(consensus_config) = validator_config.consensus_config.as_mut() {
            consensus_config.consensus_address = Multiaddr::empty();
            consensus_config.consensus_db_path = PathBuf::from("/tmp/foo/");
            consensus_config
                .narwhal_config
                .consensus_api_grpc
                .socket_addr = Multiaddr::empty();
            consensus_config
                .narwhal_config
                .prometheus_metrics
                .socket_addr = metrics_addr
        }
    }
    assert_yaml_snapshot!(network_config, {
        ".genesis" => "[fake genesis]",
        ".validator_configs[].genesis.genesis" => "[fake genesis]",
    });
}
