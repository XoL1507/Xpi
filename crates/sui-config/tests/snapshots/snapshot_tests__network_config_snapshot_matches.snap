---
source: crates/sui-config/tests/snapshot_tests.rs
expression: network_config
---
validator_configs:
  - protocol-key-pair: m/SaagdV+VOBH84SXyaD1QQpw7tJ4HQUfgCJpS6uFV+5xu4WMO8+cRFEpkjbBruyKE9ydM++5T/87lA8waSSAA==
    account-key-pair: AK7z9KS47KHfw0M2G/jkNr1C3pJZwEuDFOuOIFTdboKrBWT4edJ648As6Cg0rPqMeTpinyyg3mkZYQvoL0ETJr4=
    network-key-pair: AIpXYuIawc2zhwRCx3tMOvWMfO24d50CcObU8eL3Nn10C9WIQSA+dP6G/HEzjOAXPcYo67cZvcvMFRWFIUzAibQ=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
  - protocol-key-pair: QiWNzaFM8RHGAriXG4zIQ+keRsqQUVHAJ0SmsBfmkxbMYjMuNLstXNafYO+7KjbLkWx+tFgwHqNmNsTbsBK9iA==
    account-key-pair: ACOO227pqR/cjnG+UMCj+zzyktkCTdusAqqnCQlrCqnasgzWfEvezFOOi+mQwbZCXWi/06b+l2k+SEY1FZbMqKs=
    network-key-pair: AJ1HYWoY3yA5MH8GQiJgJYi1Kmquqntv6Ma02euop4ex9Z/d0Lf1LcwMYKRIy/lRFhCwp0Lx5NI4p6RcrgVOwhw=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
  - protocol-key-pair: 4BB+FmG6xjjQpGNs/bv21foY4nMY2jMCltCjhxgNbIzqP8527hCzpkX7MEBRHfq0yc1gWPQZKQFNFkZvXjssKg==
    account-key-pair: AK+0kF6VUM9GKoIAeCSgVdl1rpNEcMR32d27bSsmTDb60gL1Yotsq+ENWrgyROSyMO/U3MzvxGNYktNenctyJd8=
    network-key-pair: AJfn7YQ/qxpCzpwhKUNSHsIqAlaHR8wAO4gfvSf3C9/g9Lcs30s8QZpYQnBLH+55QCOl6uAiOVb7LEcBbhB7hE4=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
  - protocol-key-pair: kYP0d2s+pZoXgeVko+rpZ2k5e4q0NbJd2kGJ4QPQMc9cR/734xE9wBxW701He/5RV6/DOWmgFPi7VTfhjM7GTA==
    account-key-pair: AIM52ZwzO7Oid/63gLjUbKwwzc2QUNviNfuMxnuRwbBqnZTXHAccZRpZcVXhUvvQe7WAB32Ak1vyNmNCX/Kl8ak=
    network-key-pair: AIZBl/TMXKtsjt1LecSF77mlVXGLdbYD4ui5j1XsDeJnpzzqy7X725Nhn+RxRHmPNmU97IGS14ftQFbEczxPAHY=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
  - protocol-key-pair: ZyDgEy8FWgoN24poviNcDhIx9n6fKOuCHqxbHTnJPg+9UN97gnKsq+V1YAwCB2p5S8AIWrDXl7g9SiPgHI58JA==
    account-key-pair: APqCqu1RU9KQEL45bxLYtmFMOXx1xGE/27NtT/cE5QCrEQwhVD8hX7efjIVAYfuALmRyfmPqQ4ebKA5PameIh5A=
    network-key-pair: ABbIisgBp4THAoTQX/i9gfBB1+pjLqfKKifGyv/WUZMM+K6cRt6kc+jMcFpq8yJS6MNJtJTF3O/XGG0atPFaJY0=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
  - protocol-key-pair: RuBQj94iWraVKHboVHWawvBtJms/9hWcJ38WR6jgPck7oQxlS+QUbas9bauAs84kzjfbfHRcKyUyBdYvlnsAsw==
    account-key-pair: AHbeP6B62JU1oXZrN5h1fIqzrvW+h97hbp4Qbdi1CXKePfZvwVtLB44RZJaahQ4x3/MIjSqeJ7+3cZ5iHO/Lu9o=
    network-key-pair: AJa4rGKhGnR+XIC4bE9RFuaurlSscAly0yUDt5JWoVMvNeqo634LMrcrpZYZ2dT5jOG8wyg5tK6nYrRGYFGa4WE=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
  - protocol-key-pair: a7tE8gojj/hBeXcgQGHOCDL2HfJ2KPJ4Q1ggrXvO/+ZVrIhjM+yKuhWtjh6KURRUoIz8t81wxY3YQsqgcI0avw==
    account-key-pair: ACrfZkZbhZAbp8H6tlUREGwV7kmBjrBaPeWwPzyK71YXJfcI7/BU6ez/pZl1vtJqtInUc598clMt9PH7zXa0JpA=
    network-key-pair: AJJuq7ooTMPxmJ1zkoKcpA7WW6s+2BhjqWrbPms9/1+kwLPaTXduKbN+3+E0nSv8HiQl8vlvWvap8+WWhpfrxhw=
    db-path: /tmp/foo/
    network-address: ""
    json-rpc-address: "0.0.0.0:1"
    websocket-address: ~
    metrics-address: "0.0.0.0:1"
    admin-interface-port: 8888
    consensus-config:
      consensus-address: ""
      consensus-db-path: /tmp/foo/
      narwhal-config:
        header_size: 1000
        max_header_delay: 100ms
        gc_depth: 50
        sync_retry_delay: 5000ms
        sync_retry_nodes: 3
        batch_size: 500000
        max_batch_delay: 100ms
        block_synchronizer:
          certificates_synchronize_timeout: 2000ms
          payload_synchronize_timeout: 2000ms
          payload_availability_timeout: 2000ms
          handler_certificate_deliver_timeout: 2000ms
        consensus_api_grpc:
          socket_addr: ""
          get_collections_timeout: 5000ms
          remove_collections_timeout: 5000ms
        max_concurrent_requests: 500000
        prometheus_metrics:
          socket_addr: /ip4/127.0.0.1/tcp/1234
    enable-event-processing: false
    enable-gossip: true
    enable-checkpoint: true
    enable-reconfig: false
    grpc-load-shed: ~
    grpc-concurrency-limit: 20000000000
    genesis:
      genesis: "[fake genesis]"
account_keys:
  - SmwxHcV2gQzxKoBcs5t39VDlpi6qn4CTFlvWdcQ2bo98EQEujhdX27EEV3Z86OoMCQvPAkniwEgO0RrLK3A7sA==
  - LskdaKDI6rfXAXjiEd/Ot3tQDPgwFi2lyCfU68jIPj+LhW8+xQtCFKb75H6GmzNpzkQ4ucluHTY0AxJw//j4vQ==
  - Q9L5PWy1M9rnj1FyqfjKGyV3jMVL8RAzMJw9OdIifHVyx5W5GuC+jT9DRonmv0ccpDJ/8SX6AXiamavFW0KA5g==
  - UzKENORv6NhWOakmfPhCH8ohVHr6Ydy2XxsoQzbtSf7RwP29fVH7wkswAtz/hkGYwp98tdl70TwMB75z9G9NRw==
  - zqXFxYQpWyrIZx3ICuKRy6ie+Wm8m95nm9W63uaig36qOKjHsxwR6y1zCu5mSIno9fisAX34MT7/Vna0Cbrs4A==
genesis: "[fake genesis]"

